Dozuki/CloudPrem-Infra
# Grafana MySQL Persistance & General BI Refactor

# Problem
When we upgrade a user it's possible it will require a replacement of the Grafana helm chart. Grafana should be treated as ephemeral, however if the helm chart is deleted so is its PVC storing the customers custom configuration requiring a lengthy manual migration process.

# Solution
We leverage Grafana's ability to use a MySQL database as it's storage backend instead of the PVC. This allows the Grafana pod and helm release in general to be treated as ephemeral and can come and go as necessary to free us up for upgrades or refactors to the infra. 

## Caveats & Additional Changes

### Migration Required
We will need to manually migrate all BI customers to MySQL from SQLite when we run this upgrade. There is already a github project created by the Grafana devs for this purpose as this is a common use-case. But it must be scheduled.

### Still creating a PVC
To maintain backwards compatibility I have not removed the helm configuration to create the PVC for Grafana as this would trigger a delete of the PVC before we had a chance to migrate them to MySQL. Once all customers have been moved to MySQL we can safely remove the PVC config.

### Security Changes
I've tweaked a lot of the knobs of the Grafana config to make it as secure as possible by disabling many of the services it usually provides that we have no need for. You can see the changes in the `data` block of the `logical/bi.tf` file. Most notably disabling the metrics server, and enabling secure cookies and XSS protections. (I've also enabled `samesite` which restricts loading in an iframe so if we decide to go back to that within the app then this setting will need to be disabled)

### Misc Changes

* Grafana username is now configurable. It will use the `var.customer` value as the username or `dozuki` if none is provided.
* Refactored all the BI resources in the `physical/rds.tf` file into it's own `physical/bi.tf` file as that rds file was getting mighty large. 
* Using a k8 job to create the Grafana db along with the necessary configmap for the script and secret for the mysql credentials
* Using a k8 job to wait for the ingress controller to be ready before we deploy Grafana since the helm chart creates its own ingress object and attempting to add an ingress object before the controller is ready triggers a validation error and prevents creation. In the future we may want to look into consolidating our ingress creation in one place to prevent this race condition.
* Disabled use of `engine_version` for the DMS replication instance. For some reason terraform allows you to specify an engine version while also specifying `allow_minor_upgrade` but then when AWS upgrades the instance automatically terraform tries to downgrade to the specified version in the file.... so anyway we are not specifying a version number anymore and will default to the latest.
* Proper IAM roles for DMS CloudWatch are now created so we can get logs on the DMS job to debug issues.

## QA
You should be able to deploy this branch on any of our organizational AWS accounts. If you want to test the deploy on govcloud, be sure to set the `external_fqdn` variable to the CNAME record you manually created . 

As usual I have validated that we can safely upgrade from what's currently in master (the LE stuff) to this branch.
